amiadapters/storage/snowflake.py

```diff
@@ -0,0 +1,188 @@
+from datetime import datetime
+import json
+from typing import List
+import pytz
+
+import snowflake.connector
+
+from amiadapters.base import GeneralMeter, GeneralMeterRead
+from amiadapters.config import AMIAdapterConfiguration
+from amiadapters.storage.base import BaseAMIStorageSink
+
+
+class SnowflakeStorageSink(BaseAMIStorageSink):
+    """
+    AMI Storage Sink for Snowflake database. Implementors must specify how to store
+    raw data in Snowflake.
+    """
+
+    def __init__(self, transformed_meter_file: str, transformed_reads_file: str):
+        super().__init__(transformed_meter_file, transformed_reads_file)
+
+    def store_transformed(self, config: AMIAdapterConfiguration):
+        with open(self.transformed_meter_file, "r") as f:
+            text = f.read()
+            meters = [GeneralMeter(**json.loads(d)) for d in text.strip().split("\n")]
+
+        with open(self.transformed_reads_file, "r") as f:
+            text = f.read()
+            reads = [
+                GeneralMeterRead(**json.loads(d)) for d in text.strip().split("\n")
+            ]
+
+        conn = snowflake.connector.connect(
+            account=config.snowflake_account,
+            user=config.snowflake_user,
+            password=config.snowflake_password,
+            warehouse=config.snowflake_warehouse,
+            database=config.snowflake_database,
+            schema=config.snowflake_schema,
+            role=config.snowflake_role,
+            paramstyle="qmark",
+        )
+
+        self._upsert_meters(meters, conn)
+        self._upsert_reads(reads, conn)
+
+    def _upsert_meters(self, meters: List[GeneralMeter], conn):
+
+        create_temp_table_sql = (
+            "CREATE OR REPLACE TEMPORARY TABLE TEMP_METERS LIKE METERS;"
+        )
+        conn.cursor().execute(create_temp_table_sql)
+
+        row_active_from = datetime.now(tz=pytz.UTC)
+
+        insert_to_temp_table_sql = """
+            INSERT INTO temp_meters (
+                org_id, device_id, account_id, location_id, meter_id, 
+                endpoint_id, meter_install_date, meter_size, meter_manufacturer, 
+                multiplier, location_address, location_city, location_state, location_zip,
+                row_active_from
+            ) 
+            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
+        """
+        rows = [self._meter_tuple(m, row_active_from) for m in meters]
+        conn.cursor().executemany(insert_to_temp_table_sql, rows)
+
+        # We use a Type 2 Slowly Changing Dimension pattern for our meters table
+        # Our implementation follows a pattern in this blog post: https://medium.com/@amit-jsr/implementing-scd2-in-snowflake-slowly-changing-dimension-type-2-7ff793647150
+        # It's extremely important that the source table has no duplicates on the "merge_key"!
+        # Also, if new columns are added, be careful to update this query in each place column names are referenced.
+        merge_sql = f"""
```

Comment:
I had to do something a little funky in Snowflake to get our Slowly Changing Dimension pattern to work. Remember, this pattern lets us store historical rows for meters when they change. The `row_active_until` value tells us if a row is old or active. The "on match update old row and also insert new row" feature you'd want here doesn't quite exist in Snowflake, so I found this strategy on the internet.